File: C/code/linkedList.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


typedef struct ListNode_ {
	struct ListNode_ *next;
	int val;
} ListNode_t;

const int nodeSize = sizeof(ListNode_t);


ListNode_t *reverseList (ListNode_t *head) {

	ListNode_t *prev = NULL;
	ListNode_t *node = head;

	while (node) {
		ListNode_t *tmp = node->next;
		node->next = prev;
		prev = node;
		node = tmp;
	}
	return prev;
}

void printList (ListNode_t *node){
	while (node) {
		printf("\n %d ", node->val);
		node = node->next;
	}
}

int insertNodeList (ListNode_t **head, int val) {
	ListNode_t *node = (ListNode_t *)malloc(nodeSize);
	if (!node) {
		return -1;
	}

	memset(node, 0, nodeSize);
	node->val = val;

	node->next = *head;
	*head = node;
	return 0;
}

ListNode_t *buildList(int nums[], int count) {

	int i;
	ListNode_t *head=NULL;

	for (i=0; i < count; i++) {
		insertNodeList(&head, nums[i]);
	}

	return head;
}

int main (int argc, char *args[]) {

	int ar [] = {1};
	ListNode_t *head = buildList(ar, 1);
	printList(head);
	head = reverseList(NULL);
	printList(head);
}

File: GO/code/LinkedList.go
package main

import (
	"fmt"
)

type Node struct {
	item int
	next *Node
}

type LinkedList interface {
	Create () *interface{}
	Add (interface{}) *interface{}
	Delete(interface{}) *interface{}
}

func (Node) Create() *Node {
	var head *Node = nil
	return head
}

func (head *Node) Add (i int) *Node {
	var tmp  = new(Node)
	tmp.item = i
	tmp.next = head
	return tmp
}

func main() {
	
	

	var head *Node = nil

	for i:=0; i < 5; i++ {
		if head == nil {
			head = new(Node)
			head.item = i
			head.next = nil
		} else {
			tmp := new(Node)
			tmp.item = i
			tmp.next = head
			head = tmp
		}
	}

	head = head.Add(6)
	var node = head
	for ;node != nil;  {
		fmt.Print(node.item, " ")
		node = node.next
	}

	var l LinkedList = *head

	l = l.Add(7)
	node = l
	for ;node != nil;  {
		fmt.Print(node.item, " ")
		node = node.next
	}
}
